### ISSUE TITLE
Slow get for config false leaf

### ISSUE THREAD

**Issue by gitter-runner** [sysrepo/sysrepo#1981]
Hi,
I use _ietf-interfaces_ with several complex augments and I need to get some leaves periodically from _sysrepo_.
I experienced that the CPU usage is getting higher with increasing number of instances.

After investigation I found that getting a single 'config false' leaf from operational datastore can use
a lot of CPU.

I created a simple example to demonstrate it and attached the needed files.
The example uses the following standard models:

ietf-interfaces@2018-02-20
ietf-microwave-radio-link@2018-10-03
ieee802-ethernet-interface@2018-05-31
iana-if-type@2018-07-03


With one instance it is fast to get oper-status (measured on x86):

```
$ ./subscribe &

 ========== LISTENING FOR REQUESTS ==========

$ ./create_eth 1
Creating eth0
OK

$ time ./sr_get_items_example "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status" operational
Application will get "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status" from "operational" datastore.

/ietf-interfaces:interfaces/interface[name='eth0']/oper-status = up

real	0m0,069s
user	0m0,052s
sys	0m0,016s
```


When there are 200 instances it takes about half second of CPU time:

```
$ ./create_eth 200
...
OK

$ time ./sr_get_items_example "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status" operational
Application will get "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status" from "operational" datastore.

/ietf-interfaces:interfaces/interface[name='eth0']/oper-status = up

real	0m0,578s
user	0m0,561s
sys	0m0,017s
```


It gets worse when there are further, complex augments for interfaces list.
On my target device with slower CPU and more augments it takes about one second of CPU time to get a leaf from operational when there are 50 instances.

Based on perf measurement the function _sr_module_oper_data_add_state_default_ uses the most time (x86):
   - main                                                                                                                                                                                                         
      - 88,12% sr_get_items                                                                                                                                                                                       
         - 87,93% sr_modinfo_data_load                                                                                                                                                                            
            - sr_modinfo_module_data_load                                                                                                                                                                         
               - 87,30% sr_module_oper_data_dup_enabled                                                                                                                                                           
                  - 86,99% sr_module_oper_data_add_state_default                                                                                                                                                  
                     + 73,50% lyd_find_path                                                                                                                                                                       
                     + 13,44% lyd_validate_modules     

On the target device with more augments the _lyd_validate_modules_ uses the most time.

Can you help me how can I solve this performance bottleneck?
Is it possible to have a cache for operational similar to _SR_CONN_CACHE_RUNNING_?

[example.zip](https://github.com/sysrepo/sysrepo/files/4634789/example.zip)

libyang version: 1.0.168
sysrepo version: 1.4.60

Thanks!

**Comment by michalvasko** [comment]
Hi,
firstly I must disappoint you because no optimizations will be done in *sysrepo* until it uses the new *libyang*. We are currently working on that and we want to have it finished by the end of June. Then it will take at least a couple of weeks for *sysrepo* to migrate to it.

Also, when that is done I believe the problem you encountered will incidentally not be present anymore. One of the reasons we are rewriting *libyang* is because during new *sysrepo* implementation it became obvious that its API is quite lacking/limited. That is actually at fault in this use-case because I did what I could with the existing API so it was often utilized in ways it was not designed to be used. That could result in great inefficiencies such as this one because it has a ridiculous complexity of *O(n^2)* where *n* is the number of `interface` instances.

TLDR; You will have to wait several months before this is fixed. Please focus on correctness if you want to help us test *sysrepo* because that should be fully covered, but wait with any performance tests. Also, I will keep this use-case because while being trivial, it illustrates the problem in the current *sysrepo* and its complexity well so I will use it later to check that it works as it should.

Finally, to give you some hints if you really need this particular use-case improved, I believe you could try using *push* operational data subscription instead of the callback *pull* one because the root of the problem is in default values being added to the data after each callback called for every interface. For *push* operational data this is performed only once and then stored (so it actually works a bit like a cache, which, by the way, does not make sense for operational data in general). However, with *push* operational data you need to handle interface additions and add the `oper-status` for each of them when they are created.

Regards,
Michal

**Comment by gitter-runner** [comment]
Hi,

Could you share the progress of the new libyang?

Thanks!

**Comment by michalvasko** [comment]
Hi,
yes, the API is finished and we are in the final stages of testing/polishing it. The migration of *sysrepo* to the new *libyang* should commence in a few weeks.

Regards,
Michal

**Comment by awesomenode** [comment]
Hi,

I attached a new test application with complex augments.
[test.zip](https://github.com/sysrepo/sysrepo/files/6560868/test.zip)

-DLIBYANG2 option is needed for C files if they are compiled with libyang2/sysrepo2.

**libyang1:**

```
$ ./install.sh
$ sysrepocfg --import=config.xml -d startup
$ ./subscribe &
$ time ./test 200
Repeating test 200 times

real 0m8,320s
user 0m8,238s
sys 0m0,048s
```


**libyang2:**

```
$ ./install.sh
$ sysrepocfg --import=config.xml -d startup
$ ./subscribe &
$ time ./test 200
Repeating test 200 times

real 0m11,693s
user 0m11,543s
sys 0m0,085s
```

The following versions were tested:

```
libyang1 version: 1.0.233
sysrepo1 version: 1.4.134

libyang2 version: ce603720395db48f02966554d0a4306d4f98b3f5
sysrepo2 version: 07f0aa8
```

**Comment by michalvasko** [comment]
I finally got to some optimizations and I was able to do 2 major ones that should make this use-case ~3 times faster (so faster than *libyang* v1). Note that there are also already planned optimizations in *libyang* in regards to the LYB data format and its parsing (which is currently a bit slower than in the old *libyang*). These should be implemented in the following weeks and transparently make this use-case faster (because currently LYB parsing takes 50% of the time needed for this use-case), I am guessing something like 10 - 20% overall.
